### PR TITLE
Add builder to serve a SPA on SKY UX Host

### PR DESCRIPTION
**Pull request by Blackbaud-SteveBrush**
- Launches an Angular CLI application using the SKY UX Host URL.

### Test it locally:
- Run `npm install` and `npm start` to transpile the files.
- Change directory to `/builders-test-app` and run `npm install` and `ng serve`. This will launch the URL using the SKY UX Host URL.
- To launch the application using localhost only, run `ng serve --skyux-launch local`.

### CODE REVIEW

File: src/builders/browser/browser-options.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import {
+  BrowserBuilderOptions
+} from '@angular-devkit/build-angular';
+
+export type SkyuxBrowserBuilderOptions = BrowserBuilderOptions & { };
```
Comment (Blackbaud-SteveBrush) [inline src/builders/browser/browser-options.ts:5]: This is a placeholder type for any future `browser` builder options.

File: src/builders/dev-server/dev-server-options.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+export type SkyuxDevServerBuilderOptions = DevServerBuilderOptions & {
+
+  skyuxHostUrl?: string;
+
+  skyuxLocalUrl?: string;
+
+  skyuxOpen?: 'host' | 'local';
+
+  skyuxOpenBrowser?: string | string[];
```
Comment (Blackbaud-SteveBrush) [inline src/builders/dev-server/dev-server-options.ts:13]: ☝️ @Blackbaud-PaulCrowder From my research, it doesn't look like Angular CLI supports writing complex objects as CLI arguments (e.g. `ng serve --skyux.hostUrl` for `{ skyux: { hostUrl: string } }`). Today, they only support `boolean`, `number`, and `string` types, so we'll probably need to stick with "flat" property names, at least for the options that we want to provide to the command line.

File: src/builders/dev-server/schema.ext.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    "skyuxOpen": {
+      "type": "string",
+      "description": "Opens URL in localhost or with the SKY UX host URL.",
+      "enum": [
+        "local",
+        "host"
+      ],
+      "default": "host"
+    },
+    "skyuxOpenBrowser": {
```
Comment (Blackbaud-SteveBrush) [inline src/builders/dev-server/schema.ext.json:21]: @Blackbaud-BobbyEarl Just confirming: none of these properties will be used for `ng build`?

File: src/webpack/plugins/save-metadata.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    addAssetSourceTap(
+      PLUGIN_NAME,
+      compiler,
+      (content, file) => `${content}\nvar ${getFallbackName(file)} = true;`
+    );
+
+    compiler.hooks.done.tap(PLUGIN_NAME, (webpackStats) => {
+
+      const stats = webpackStats.toJson();
+      const assets = getSortedAssets(stats, true);
```
Comment (Blackbaud-SteveBrush) [inline src/webpack/plugins/save-metadata.ts:30]: This only returns entry-level chunks (e.g. 'main.js', 'polyfills.js', etc.). I'm assuming we'll need to include all of the lazy-loaded module chunks in the `metadata.json` file too? Should that be included in a separate pull request?

File: src/builders/dev-server/dev-server-options.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import {
+  homedir
+} from 'os';
+
+/**
+ * The input options passed to the builder.
+ */
+export type SkyuxDevServerBuilderOptions = DevServerBuilderOptions & {
+
+  skyuxHostUrl?: string;
```
Comment (Blackbaud-SteveBrush) [inline src/builders/dev-server/dev-server-options.ts:14]: ☝️ From my research, Angular CLI does not support writing complex objects as CLI arguments (e.g. `ng serve --skyux.hostUrl` for `{ skyux: { hostUrl: string } }`). Today, they only support `boolean`, `number`, and `string` types, so we'll probably need to stick with "flat" property names, at least for the options that we want to provide to the command line.

File: src/builders/dev-server/schema.ext.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      "description": "Host URL"
+    },
+    "skyuxLaunch": {
+      "type": "string",
+      "description": "Opens URL in localhost or with the SKY UX host URL.",
+      "enum": [
+        "local",
+        "host"
+      ],
+      "default": "host"
```
Comment (Blackbaud-BobbyEarl) [inline src/builders/dev-server/schema.ext.json:15]: I think this is an opportunity for us make the differences between internal and external development better.

Instead of defaulting to `host` here, could we leave this as `local` or maybe even blank (to mirror NG CLI)?

Then, in a future PR, prompt the user in the `ng-add` schematic "Are you a Blackbaud Developer?" If so, write `skyuxLaunch: "host"` in their project's `angular.json`?

File: src/webpack/stats-utils.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  name: string;
+}
+
+export function getAssets(stats: Stats.ToJsonOutput): Asset[] {
+  if (!stats) {
+    return [];
+  }
+
+  const assets: Asset[] = [];
+  stats.chunks?.forEach(chunk => {
```
Comment (Blackbaud-BobbyEarl) [inline src/webpack/stats-utils.ts:15]: Could you use the same technique to check for `stats` and remove the `if` block above?

`stats?.chunks?.forEach(...`

File: src/webpack/plugins/open-host-url.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  Compiler
+} from 'webpack';
+
+import {
+  getAssets
+} from '../stats-utils';
+
+const PLUGIN_NAME = 'open-skyux-host-plugin';
+
+interface SkyuxOpenHostURLPluginConfig {
```
Comment (Blackbaud-BobbyEarl) [inline src/webpack/plugins/open-host-url.ts:13]: For development, wouldn't we want this exported in its own file so it can be referenced?

There are additional properties, that individual SPAs will eventually need to be able to set here as well - assumingely from their `angular.json` file.

File: src/webpack/stats-utils.ts (hunk 7 above)
Comment (Blackbaud-SteveBrush) [reply src/webpack/stats-utils.ts:15]: Good idea.

File: src/builders/dev-server/dev-server-options.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+} from '@angular-devkit/build-angular';
+
+import {
+  homedir
+} from 'os';
+
+/**
+ * The input options passed to the builder.
+ */
+export type SkyuxDevServerBuilderOptions = DevServerBuilderOptions & {
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/builders/dev-server/dev-server-options.ts:12]: Good idea.

File: src/builders/dev-server/schema.ext.json (hunk 6 above)
Comment (Blackbaud-SteveBrush) [reply src/builders/dev-server/schema.ext.json:15]: I was under the assumption that the builder’s default settings would lean toward an internal Blackbaud developer (i.e., launching via SKY UX Host would be the default setup). Is the idea that if a consumer installed the builder (without the initial prompt that we’ve discussed adding) it would essentially just call through to the default Angular builders and “do nothing”? Are there any features of our custom builder that would benefit an external consumer of SKY UX? If our builder simply runs Angular builders by default, I feel like we shouldn’t advertise it to public-facing consumers at all, or create a separate builder that is isolated to external folks. What do you think?

File: src/webpack/plugins/open-host-url.ts (hunk 8 above)
Comment (Blackbaud-SteveBrush) [reply src/webpack/plugins/open-host-url.ts:13]: Good thought.

File: src/builders/dev-server/schema.ext.json (hunk 6 above)
Comment (Blackbaud-BobbyEarl) [reply src/builders/dev-server/schema.ext.json:15]: Good question... sounds like we're taking this discussion internally for now.  :-)

**Review by Blackbaud-BobbyEarl — APPROVED** [review]
